### DISCUSSION TITLE
freeze

### DISCUSSION THREAD

**Discussion by mauroForlimpopoli** [Moddable-OpenSource/moddable#658] [Q&A]
I done a scroller. This is the code:
```js
const ScrollerAct = Label.template($ => ({
    anchor: "BackGround", top: 15, bottom: 15, left: 0, right: 0,
    skin: new Skin({ fill: TRANSPARENT }), style: new orange28(),
    string: $, name: $,
	active: true,
	Behavior: ScrollerActBehavior
}));
Object.freeze(ScrollerAct);
const scrollerAllarm = new Scroller(null, {  
	left: 120, right: 0, top: 0, bottom: 0, 
	skin: new Skin({ fill: MOZILLA_BLUE }),
	active: true, backgroundTouch: true, clip: true,
	Behavior: VerticalScrollerBehavior,
	contents:[
		Column(null, {
			top: 0, left: 20,
			contents: [
				new ScrollerAct("Canale 1",{name:100}),
				new ScrollerAct("Canale 2",{name:101}),
				new ScrollerAct("Canale 3",{name:102}),
				new ScrollerAct("Canale 4",{name:103})
			]
		})
	],
});	
Object.freeze(scrollerAllarm);
```
I freezed the VerticalScrollerBehavior and ScrollerActBehavior, 
Can I freeze it or preload the scroller like I done ? 
Regards
regards

**Comment by lprader** [comment]
The questions "can I freeze x" and "can I preload x" are very different questions :) Freezing is something you do to JavaScript objects. Preloading is something you do to a JavaScript module.

Correct me if I'm wrong, but it looks like you've already determined that you can freeze the `scrollerAllarm` object, and you are asking if you can preload the module that instantiates the `scrollerAllarm` object.

To answer that question: no, you can't preload a module that instantiates Piu objects (like scrollers, labels, styles, etc.). For example, say your `main.js` module consists of the following code. You cannot preload this module, because it instantiates a Piu `Content` object:

```js
const BlueBox = new Content(null, {
	top: 0, bottom: 0, left: 0, right: 0,
	skin: new Skin({ fill: "blue" })
});

export default new Application(null, { 
	contents: [
		BlueBox
	]
});
```

You can, however, write this module in a way that can be preloaded. For example, you can put the code that instantiates the `Content` object in a function that won't be called at build-time, but will be called when the app launches on the device:

```js
export default function() {
	const BlueBox = new Content(null, {
		top: 0, bottom: 0, left: 0, right: 0,
		skin: new Skin({ fill: "blue" })
	});
	return new Application(null, { 
		contents: [
			BlueBox
		]
	});
}
```

Piu [templates](https://github.com/Moddable-OpenSource/moddable/blob/public/documentation/piu/piu.md#templates) are also a useful tool that you can use to make modules preloadable. The `template` function returns a template, which is a constructor of a Piu object (rather than an instance of a Piu object).  A module can still be preloaded if it creates templates, so here's another way you could rewrite that module so that it can be preloaded:

```js
const BlueSkin = Skin.template({ fill: "blue" });

const BlueBox = Content.template($ => ({
	top: 0, bottom: 0, left: 0, right: 0,
	Skin: BlueSkin
}));

const BlueApp = Application.template($ => ({
	contents: [
		new BlueBox($)
	]
}));

export default function() {
	return new BlueApp;
}
```

So, to summarize:
- the module you shared cannot be preloaded as-is because it instantiates Piu objects
- ...but you can rewrite the module so it can be preloaded

**Comment by mauroForlimpopoli** [comment]
Dear Lizzie,
thank you for your assistance, but unfortunately something is not clear for me. 
I need to save memory because my project is near memory full.
This is my manifest:
```
{	
	"build": {
		"PARTITIONS_FILE": "./partitions.csv"
	},
	"include": [
		"$(MODDABLE)/examples/manifest_base.json",
		"$(MODDABLE)/examples/manifest_piu.json",
		"$(MODULES)/input/keyboard/manifest.json",
		"$(MODULES)/pins/analog/manifest.json",
		"$(MODDABLE)/modules/files/file/manifest.json",
		"$(MODULES)/network/wifi/manifest.json",
		"$(MODDABLE)/examples/manifest_net.json",
		"$(MODDABLE)/modules/network/ble/manifest_server.json",
		"$(MODULES)/network/mdns/manifest.json",
		"$(MODDABLE)/modules/io/manifest.json",
		"$(MODULES)/pins/digital/manifest.json"
	],
	"creation":{
		"stack":510,
		 "static": 98304,
		 "keys": {
			"available":30
		},
		"chunk": {
			"initial":35362,
			"incremental":1024
		}
	},
	"modules": {
		"*": [
			"./main",
			"./assets",
			"$(MODULES)/network/http/*",
			"./funcSleep",
			"$(MODULES)/network/sntp/*",
			"$(MODULES)/drivers/onewire/*"
		],
		"esp32/ota": "$(MODDABLE)/build/devices/esp32/modules/ota/*"
	},
	"resources":{
		"*-alpha": [
			"$(MODDABLE)/examples/assets/fonts/OpenSans-Regular-20",
			"$(MODDABLE)/examples/assets/fonts/OpenSans-Semibold-18",
			"$(MODDABLE)/examples/assets/fonts/OpenSans-Light-42",
			"$(MODDABLE)/examples/assets/fonts/OpenSans-Semibold-28",
			"$(MODDABLE)/examples/assets/fonts/OpenSans-Regular-72",
			"$(MODDABLE)/examples/assets/fonts/arc64",
			"$(MODDABLE)/examples/assets/fonts/roboto32"
		],
		"*-image(50)": [
			"$(MODDABLE)/examples/assets/images/Bluetooth-icon",
			"$(MODDABLE)/examples/assets/images/Wifi-icon",
			"$(MODDABLE)/examples/assets/images/unlock-icon",
			"$(MODDABLE)/examples/assets/images/lock-icon",
			"$(MODDABLE)/examples/assets/images/down-icon",
			"$(MODDABLE)/examples/assets/images/battery-icon"
		],
		"*-mask": [
			"$(MODDABLE)/examples/assets/images/loading-icon",
			"$(MODDABLE)/examples/assets/images/thermometer-mask",
			"$(MODDABLE)/examples/assets/images/battery-mask"		
		]
	},
	"platforms": {
		"esp32": {
			"config": {
				"rotation": 90,
				"brightness": 0,
				"file": {
					"root": "/mod/"
				},
				"onewire": {
					"pin": "16"
				}
			},
			"defines": {
				"onewire": {
					"driver_rmt": "RMT",
					"rmt_rx_channel": "RMT_CHANNEL_2",
					"rmt_tx_channel": "RMT_CHANNEL_3"
				}
			},
			"include": [
				"$(MODULES)/pins/digital/manifest.json"
			],
			"modules": {
				"*": [
				"$(MODULES)/files/flash/*",
				"$(MODULES)/files/flash/esp32/*",
				"$(MODULES)/drivers/onewire/esp/*"
				]
			}
		}
	},
	"preload": [
		"neopixel",
		"http",
		"sntp",
		"file",
		"pins/analog",
		"timer",
		"esp32/ota",
		"ble",
		"preference",
		"flash",
		"Resource",
		"keyboard",
		"time",
		"timer",
		"onewire",
		"DS18X20",
		"wifi",
		"net",
		"mdns",
		"assets"
	],
	 "data": {
        "*": [
            "./mydata",
			"./myWeb"			
        ]
    },"ble":{
		"*": [
			"./bleservices/*"
		]
	}
}
```
----
I have a questions
During running the RAM Free is about 46000 kb and when I active bluetooth server with ( let server = new UARTServer; ) it goes down to 9700 kb.
So I wanted to preload the BLE Server, but I don't see variation on my RAM usage. Maybe did I something wrong ?
To follow your indications I also tried to implement a assets.js , but when I add "import ASSETS from "assets";"
The system goes in full memory
What can I do?
Thank you

**Reply by lprader** [reply]
If you look at the [BLE server manifest](https://github.com/Moddable-OpenSource/moddable/blob/public/modules/network/ble/manifest_server.json#L15) that you have in your `includes` array, you can see that it preloads all of the BLE server modules already. There's no need to do it in your app manifest too.

Regarding your memory issues, there are many techniques you can use to reduce memory use, but it's impossible for me to make recommendations just by looking at your manifest and a snippet of scroller code. If you have more specific questions, I can try to answer :)

Also, a formatting tip: it would be **very** helpful if you put your code in code blocks. You can do that by surrounding code with three backticks, like this:

    ```
    code goes here
    ```

**Reply by phoddie** [reply]
> During running the RAM Free is about 46,000 bytes and when I active bluetooth server with ( let server = new UARTServer; ) it goes down to 9,700 bytes. So I wanted to preload the BLE Server, but I don't see variation on my RAM usage.

The BLE implementation in ESP-IDF Is NimBLE.  It uses about 30 KB of RAM when active. Preload can reduce the memory use of JavaScript code, however it does not help with native code like NimBLE. 

FYI - The ESP-IDF (and consequently thee Moddable SDK) and previously used Bluedroid for BLE. Bluedroid uses much more memory than NimBLE, so the current memory use is an improvement.

**Reply by mauroForlimpopoli** [reply]
Unfortunately the memory architecture is still not very clear to me. I have a "memory full" problem and looking the XBug I think it depends on the slot used. At the start of the program I have used 35kB of my 50 kB available. To reduce this one I tried to preload some modules.
So I done these changes:
```
const KeyboardApp = Application.template($ => ({
	active: true, Skin: WhiteSkin,
	contents: [
		new SampleContent($),
	],
	labels: [
		new ScrollerAct($),
		new ScrollerKb($),
		new menuLabelOne($)		
	],
	colums: [
		new KeyboardContainer($)
	],
	Behavior: class extends Behavior {
		onCreate(application, data) {
			application.add(bgBlu);
			application.add(startForm);
		}
	}
}));
```
and I this is a piece of my manifest to preload them:
```
	"preload": [
		"neopixel",
		"http",
		"sntp",
		"file",
		"pins/analog",
		"timer",
		"esp32/ota",
		"ble",
		"preference",
		"flash",
		"Resource",
		"keyboard",
		"time",
		"timer",
		"onewire",
		"DS18X20",
		"funcSleep",
		"wifi",
		"keyboard",
		"SampleContent",
		"ScrollerAct",
		"ScrollerKb",
		"menuLabelOne",
		"KeyboardContainer"
	],
```
But the use of slot used is increased . What did I wrong?
Furthermore I read many times the article on Preload, but I need an example on how to preload a simple function like this:
```
function decodSens(val) {
	if(val == 252) {
		val = "ON"
	} else if (val == 251) {
		val = "Off"
	} else if (val == 1) {
		val = "Out"
	} else {
		val = val + 1
		val = val / 2
		val = val - 60
	}
	return val	
}
```
Thank you

**Comment by phoddie** [comment]
@mauroForlimpopoli, you are asking complex questions but providing few details. You are assuming that preload will solve your memory problem. It might, but that depends on what is using memory. For example, a complex UI with instances of many objects requires memory and that preload cannot help with.

Some notes:

- All functions declared at the top level of a module that is preloaded are preloaded. If `decodSens` is a top level function, nothing more is needed.
- You say, "at the start of the program I have used 35kB of my 50 kB available." It isn't clear what you mean by the start of the program. For example, is it before or after any code has executed in your `main.js`?
- You say, "I tried to preload some modules". Are there any modules which are not preloaded? If yes, have any of those been loaded at the time you made your measurement of slot memory use?

**Reply by mauroForlimpopoli** [reply]
1)decodSens is a function called from a function who is  inside a class "menuLabelOneBehavior" who is freezed. Is it preloaded ?  If no is it possible to preload it?

2) Before "IMPORT" the status of the memory slot is about 3000/8000. After these steps the status is about 20000/40000.
Before "decodSens" function the state is 18672/41984 and after is 18768/41984. Generally when the program load a function che Memory slot grows. Another example is 
```
const ita = {
	txt_nik: "nikit",	
	txt_One: "One",	
	txt_info: "Info",	
	txt_display: "Display",	
	txt_bluetooth: "Bluetooth",	
	txt_update: "Update",
	txt_cleanMem: "Clear Mem.",
	txt_scansione: "Scansione",
	txt_esegLogin: "eseguire login",
	txt_sensori: "Sensori",
	txt_profilo: "Profilo ",
	txt_stampa: "Stampa",
	txt_termostato: "Termostato",
	txt_aggiorna: "Aggiorna"
}
Object.freeze(ita);
```
the slot used grow of 256 byte after this "const". is it normal?
3) "module" in the xsbug looks all preloade. The module I meant was these 
```
const ScrollerAct = Label.template($ => ({
    anchor: "BackGround", top: 15, bottom: 15, left: 0, right: 0,
    skin: new Skin({ fill: TRANSPARENT }), style: new orange28(),
    string: $, name: $,
	active: true,
	Behavior: ScrollerActBehavior
}));
```
During uploading my user interface, each time this module is called the "slot used" grows of  about 250 byte. Is it possible to prevent this growth ? I thought that by mean the previus changes (I showed before) it was preload and to avoid this growth 
Thank you

**Comment by phoddie** [comment]
You may be confusing the freezing of an object with preload. Freezing an object prevents modification of existing properties and adding new properties. That is defined by the JavaScript language. Preloading of modules is a feature of the XS JavaScript engine. It allows the body of a modules to be executed at build time on your computer, rather than at runtime on the device. Objects that are frozen during preload may be stored in flash memory, so they don't use RAM on the device.

You show examples of code that you are stepping through on the device in xsbug. That is **after** preload has finished. The code is running on the device. Objects created at runtime on the device use memory, even if you freeze them.

Consider this example you give:

```
const ita = {
	txt_nik: "nikit",	
	txt_One: "One",	
	txt_info: "Info",	
	txt_display: "Display",	
	txt_bluetooth: "Bluetooth",	
	txt_update: "Update",
	txt_cleanMem: "Clear Mem.",
	txt_scansione: "Scansione",
	txt_esegLogin: "eseguire login",
	txt_sensori: "Sensori",
	txt_profilo: "Profilo ",
	txt_stampa: "Stampa",
	txt_termostato: "Termostato",
	txt_aggiorna: "Aggiorna"
}
Object.freeze(ita);
```
You note that it is consuming 256 bytes of slots. That is expected, because you are initializing the object on the device. The `ita` objects uses two slots, plus one more slot for each of its 14 properties. Each slot is 16 bytes on an ESP32, so that is a total of (2 + 14) * 16 == 256 bytes.

To move those 256 bytes to flash memory, the `ita` object needs to be initialized in a module that is preloaded. It also needs to be deeply frozen  by calling `Object.freeze` with the optional second parameter set to true:

```
Object.freeze(ita, true);
```
